From badd2a55bcde4badc6a94ea55e46c94e70cbc311 Mon Sep 17 00:00:00 2001
From: Grzegorz Bokota <bokota+github@gmail.com>
Date: Fri, 17 Nov 2023 04:49:26 +0100
Subject: [PATCH] Fix problem with transform box of multiscale image (#6390)

fix #6382

The problem described in #6392 comes from using `extent_data` for level
0, but the Transform box visual is children of the image visual. When we
change the level of data then we update the transform matrix of the
image visual which leads to increase size of Transform Box.

In this PR I add an extent property that reflects the current data
level. So We get a smaller extent, which is fixed by scaling on the
image layer.

---------

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>
---
 napari/_vispy/overlays/base.py            |  7 ++++-
 napari/_vispy/overlays/bounding_box.py    |  2 +-
 napari/_vispy/overlays/interaction_box.py |  2 +-
 napari/layers/base/base.py                | 20 ++++++++++++++
 napari/layers/image/image.py              | 32 ++++++++++++++++++++++-
 5 files changed, 59 insertions(+), 4 deletions(-)

diff --git a/napari/_vispy/overlays/base.py b/napari/_vispy/overlays/base.py
index 61a97d4e..be2055c9 100644
--- a/napari/_vispy/overlays/base.py
+++ b/napari/_vispy/overlays/base.py
@@ -1,3 +1,5 @@
+from typing import TYPE_CHECKING
+
 from vispy.visuals.transforms import MatrixTransform, STTransform
 
 from napari._vispy.utils.gl import BLENDING_MODES
@@ -5,6 +7,9 @@ from napari.components._viewer_constants import CanvasPosition
 from napari.utils.events import disconnect_events
 from napari.utils.translations import trans
 
+if TYPE_CHECKING:
+    from napari.layers import Layer
+
 
 class VispyBaseOverlay:
     """
@@ -143,7 +148,7 @@ class VispySceneOverlay(VispyBaseOverlay):
 
 
 class LayerOverlayMixin:
-    def __init__(self, *, layer, overlay, node, parent=None) -> None:
+    def __init__(self, *, layer: "Layer", overlay, node, parent=None) -> None:
         super().__init__(
             node=node,
             overlay=overlay,
diff --git a/napari/_vispy/overlays/bounding_box.py b/napari/_vispy/overlays/bounding_box.py
index 84dbea2e..744206bc 100644
--- a/napari/_vispy/overlays/bounding_box.py
+++ b/napari/_vispy/overlays/bounding_box.py
@@ -21,7 +21,7 @@ class VispyBoundingBoxOverlay(LayerOverlayMixin, VispySceneOverlay):
         self.overlay.events.point_color.connect(self._on_point_color_change)
 
     def _on_bounds_change(self):
-        bounds = self.layer._display_bounding_box(
+        bounds = self.layer._display_bounding_box_augmented_data_level(
             self.layer._slice_input.displayed
         )
         # invert for vispy
diff --git a/napari/_vispy/overlays/interaction_box.py b/napari/_vispy/overlays/interaction_box.py
index bdf4ea84..d2f38262 100644
--- a/napari/_vispy/overlays/interaction_box.py
+++ b/napari/_vispy/overlays/interaction_box.py
@@ -67,7 +67,7 @@ class VispyTransformBoxOverlay(_VispyBoundingBoxOverlay):
 
     def _on_bounds_change(self):
         if self.layer._slice_input.ndisplay == 2:
-            bounds = self.layer._display_bounding_box(
+            bounds = self.layer._display_bounding_box_augmented_data_level(
                 self.layer._slice_input.displayed
             )
             # invert axes for vispy
diff --git a/napari/layers/base/base.py b/napari/layers/base/base.py
index c1ca5325..e2be6e5a 100644
--- a/napari/layers/base/base.py
+++ b/napari/layers/base/base.py
@@ -240,6 +240,7 @@ class Layer(KeymapProvider, MousemapProvider, ABC):
         Mode.PAN_ZOOM: 'standard',
         Mode.TRANSFORM: 'standard',
     }
+    events: EmitterGroup
 
     def __init__(
         self,
@@ -1424,6 +1425,25 @@ class Layer(KeymapProvider, MousemapProvider, ABC):
         """An axis aligned (ndisplay, 2) bounding box around the data"""
         return self._extent_data[:, dims_displayed].T
 
+    def _display_bounding_box_augmented(
+        self, dims_displayed: List[int]
+    ) -> npt.NDArray:
+        """An augmented, axis-aligned (ndisplay, 2) bounding box.
+
+        This bounding box includes the size of the layer in best resolution, including required padding
+        """
+        return self._extent_data_augmented[:, dims_displayed].T
+
+    def _display_bounding_box_augmented_data_level(
+        self, dims_displayed: List[int]
+    ) -> npt.NDArray:
+        """An augmented, axis-aligned (ndisplay, 2) bounding box.
+
+        If the layer is multiscale layer, then returns the
+        bounding box of the data at the current level
+        """
+        return self._display_bounding_box_augmented(dims_displayed)
+
     def click_plane_from_click_data(
         self,
         click_position: np.ndarray,
diff --git a/napari/layers/image/image.py b/napari/layers/image/image.py
index b908af25..9176afce 100644
--- a/napari/layers/image/image.py
+++ b/napari/layers/image/image.py
@@ -490,7 +490,28 @@ class _ImageBase(IntensityVisualizationMixin, Layer):
         extent_data : array, shape (2, D)
         """
         shape = self.level_shapes[0]
-        return np.vstack([np.zeros(len(shape)), shape])
+        return np.vstack([np.zeros(len(shape)), shape - 1])
+
+    @property
+    def _extent_data_augmented(self) -> np.ndarray:
+        extent = self._extent_data
+        return extent + [[-0.5], [+0.5]]
+
+    @property
+    def _extent_level_data(self) -> np.ndarray:
+        """Extent of layer, accounting for current multiscale level, in data coordinates.
+
+        Returns
+        -------
+        extent_data : array, shape (2, D)
+        """
+        shape = self.level_shapes[self.data_level]
+        return np.vstack([np.zeros(len(shape)), shape - 1])
+
+    @property
+    def _extent_level_data_augmented(self) -> np.ndarray:
+        extent = self._extent_level_data
+        return extent + [[-0.5], [+0.5]]
 
     @property
     def data_level(self):
@@ -1030,6 +1051,15 @@ class _ImageBase(IntensityVisualizationMixin, Layer):
         extent_at_level = np.vstack([np.zeros(len(shape)), shape - 1])
         return extent_at_level[:, dims_displayed].T
 
+    def _display_bounding_box_augmented_data_level(
+        self, dims_displayed: List[int]
+    ) -> npt.NDArray:
+        """An augmented, axis-aligned (ndisplay, 2) bounding box.
+        If the layer is multiscale layer, then returns the
+        bounding box of the data at the current level
+        """
+        return self._extent_level_data_augmented[:, dims_displayed].T
+
     # For async we add an on_chunk_loaded() method.
     if config.async_loading:
 
-- 
2.34.1

